### PR TITLE
CreateNewBlock: Child-pays-for-parent / Add transaction fee later

### PR DESCRIPTION
**Pull request by luke-jr**
**Status: Passes unit tests**

Consider parent transactions in the "cost" of child transactions until confirmed, and confirm them together

This is the part of #1240 that @gavinandresen left out of #1590 since he felt it belonged in a separate commit/pullreq.

### CODE REVIEW

**Comment by voisine** [comment]
Would love to see this get in. Have a situation right now where a new user received 0.01btc tx with no fee and then spent their entire balance, 17btc, which is stuck. It would be a real improvement for user experience if wallet software could just include a fee to cover unconfirmed input tx in addition to itself instead of trying to explain to users why they can't spend those funds yet.

**Comment by gavinandresen** [comment]
Needs tests, as suggested by gmaxwell. A regression test making sure that this would actually fix the situation described by @voisine would be excellent (although non-trivial-- do you assume that the 0.01btc tx with no fee got relayed and is still in memory pools, or not?)

And needs a good review from somebody who is good at spotting potential DoS weaknesses (like Sergio).

**Comment by ABISprotocol** [comment]
Yes, thanks to those who have worked on this so far.  This being the oldest open pull request, and having merits for inclusion in 0.11, it needs and deserves both a thoughtful review and merge.

**Comment by jgarzik** [comment]
ACK - light test at current tip

File: src/miner.cpp [hunk 1]
```diff
@@ -138,82 +276,76 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
...
+
         for (map<uint256, CTxMemPoolEntry>::iterator mi = mempool.mapTx.begin();
              mi != mempool.mapTx.end(); ++mi)
         {
             const CTransaction& tx = mi->second.GetTx();
+
+            const uint256& hash = tx.GetHash();
+            CTxInfo& txinfo = mapInfoById[hash];
+            txinfo.hash = hash;
+            txinfo.pmapInfoById = &mapInfoById;
```
Comment (sipa) [inline src/miner.cpp]: This cyclic dependency is pretty ugly. Can't you just pass the mapInfoById to the methods in txindo that need it?

**Comment by sipa** [comment]
Has anyone done benchmarks on the impact on block creation time?

File: src/miner.cpp (hunk 1 above)
Comment (luke-jr) [inline src/miner.cpp]: That's basically everything... :/

Comment (sipa) [inline src/miner.cpp]: It's easier to not put the logic in the data structure, I think. The fact that you need a pointer from the low-level data structure to the container is just a symptom of that.

File: src/miner.cpp [hunk 2]
```diff
@@ -297,22 +423,29 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
...
+                pblock->vtx.push_back(*ptxinfo->ptx);
+                pblocktemplate->vTxFees.push_back(ptxinfo->nTxFee);
+                pblocktemplate->vTxSigOps.push_back(ptxinfo->nTxSigOps);
+                nFees += ptxinfo->nTxFee;
+
+                ptxinfo->fInvalid = true;
+                if (!ptxinfo->setDependents.empty())
                 {
-                    if (!porphan->setDependsOn.empty())
+                    fResort = true;
```
Comment (morcos) [inline src/miner.cpp]: Aren't this !empty check and flag to resort redundant given the foreach loop below?

File: src/miner.cpp [hunk 3]
```diff
@@ -297,22 +423,29 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
...
 
-            // Add transactions that depend on this one to the priority queue
-            if (mapDependers.count(hash))
+            bool fResort = false;
+            BOOST_FOREACH(CTxInfo* ptxinfo, vAdded)
             {
-                BOOST_FOREACH(COrphan* porphan, mapDependers[hash])
+                pblock->vtx.push_back(*ptxinfo->ptx);
+                pblocktemplate->vTxFees.push_back(ptxinfo->nTxFee);
+                pblocktemplate->vTxSigOps.push_back(ptxinfo->nTxSigOps);
```
Comment (morcos) [inline src/miner.cpp]: I think that `ptxinfo->nTxSigOps` is just that transaction's `GetP2SHSigOpCount` and doesn't include its `GetLegacySigOpCount`.

File: src/miner.cpp (hunk 2 above)
Comment (luke-jr) [inline src/miner.cpp]: Probably better to just set fResort here and not in the foreach... (done)

File: src/miner.cpp (hunk 3 above)
Comment (luke-jr) [inline src/miner.cpp]: Good catch, fixed.

**Comment by morcos** [comment]
I've done a code review on this and it looks good to me. I've also done some performance testing in manufactured RPC tests as well as under more typical load.  It can cause CreateNewBlock to take multiples as long if there are chains of transactions in the mempool, so I do think it would be useful to test deliberately extreme chains as mentioned above.   Under typical mempool activity though it seems to be about a 40% hit, making CreateNewBlock take on average 200ms instead of 140ms.  Before and after this change the maximum time on a typical mempool is about 1 second.

**Comment by luke-jr** [comment]
@morcos If you want to push those RPC tests as a branch somewhere, I can include them when I rebase this.

**Comment by morcos** [comment]
@luke-jr Feel free to take a look at this, https://github.com/morcos/bitcoin/commits/CPFP.  I'd rebased and then added 2 commits to do some timing.  Its not really something to include as it doesn't test anything, but its just what I was using to stress the code.  I'm not quite sure why it takes so long, but I didn't look into.  See the commit text for the most recent commit.

File: src/miner.cpp [hunk 4]
```diff
@@ -138,82 +276,76 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn)
...
+                    mapInfoById[hashPrev].setDependents.insert(hash);
+                    nConf = 0;
+                }
+                else
+                {
+                    // We don't know where the input is
+                    // In this case, it's impossible to include this transaction in a block, so mark it invalid and move on
+                    txinfo.fInvalid = true;
+                    LogPrintf("priority %s invalid input %s\n", txinfo.hash.ToString().substr(0,10).c_str(), txin.prevout.hash.ToString().substr(0,10).c_str());
+                    goto nexttxn;
```
Comment (jtimon) [inline src/miner.cpp]: What about replacing this part

```
                    txinfo.fInvalid = true;
                    goto nexttxn;
                 }
                  nTotalIn += nValueIn;
             }
             nTxFee = nTotalIn - tx.GetValueOut(); 
             mempool.ApplyDeltas(hash, dPriority, nTotalIn);
             vecPriority.push_back(&txinfo); 
nexttxn:    (void)1;
```

with the following?

```
                    txinfo.fInvalid = true;
                 }
                 if (!txinfo.fInvalid)
                    nTotalIn += nValueIn;
             }
             if (!txinfo.fInvalid) {
                 nTxFee = nTotalIn - tx.GetValueOut(); 
                 mempool.ApplyDeltas(hash, dPriority, nTotalIn);
                 vecPriority.push_back(&txinfo);
             }
```

https://xkcd.com/292/

Comment (jtimon) [inline src/miner.cpp]: Or just have a separate function:

```
if (NewFunction(...)) {
    nTotalIn += nValueIn;
    nTxFee = nTotalIn - tx.GetValueOut(); 
    mempool.ApplyDeltas(hash, dPriority, nTotalIn);
    vecPriority.push_back(&txinfo);
}
```

Comment (ABISprotocol) [inline src/miner.cpp]: Does it work?

**Comment by morcos** [comment]
I think I missed a few things on my first review of this.  It seems like if B and C are both children of A, and D depends on both B and C, then the effectiveSize() of D will double count the size of A?

Also, it doesn't look like the fees of A, B, or C are counted, only the fees of the final child (D in this case).

Or am I missing something now...

**Comment by jgarzik** [comment]
I leave this pull request open in the hope that it's current status of "98% there" will get the final way there.  @luke-jr has faithfully updated it and addressed feedback over time, and consensus is that it's merge worthy.

It seems like it just needs a reviewer or two to expend a day or two of focused analysis and testing.

**Comment by sipa** [comment]
I absolutely gree on the concept of CPFP, but I believe there are various
hard to analyse edge cases with this approach.

With the work that @sdaftuar is doing for the mempool refactor (particular
the dependency tracking and limiting), we'll have a framework that is very
close to actually having a mempool indexed by dependency-inclusive
feerates, resulting a much cleaner way to implement actual CPFP mining, by
simply putting dependency packages one by one in a block, following that
sort order.

**Comment by morcos** [comment]
Its actually a fair amount of code that will still need to be written to do CPFP even after #6654, but I agree that it'll provide the right framework for implementing it.   I think the approach of starting from the dependent tracking framework and adding ancestor tracking is likely to get us to functional CPFP quicker than starting from this pull.

**Comment by luke-jr** [comment]
I agree the long-term implementation is likely to be different, but given the extreme maturity of this code (especially as opposed to the immaturity of #6654 and similar proposals), I do think it would be best to merge it now/soon, with the understanding that it may be replaced by an improved implementation later. Don't let the perfect be the enemy of the good etc.

**Comment by morcos** [comment]
@luke-jr did you see my [comment] (https://github.com/bitcoin/bitcoin/pull/1647#issuecomment-121668295) a few comments up?  I think there are a couple of things that don't work properly with the code now.

**Comment by jtimon** [comment]
I think it makes sense to try another version of #6654 on top of this if it's not too much work and take a look at that branch before merging #6654

**Comment by luke-jr** [comment]
> I think I missed a few things on my first review of this. It seems like if B and C are both children of A, and D depends on both B and C, then the effectiveSize() of D will double count the size of A?

Yes, probably. But it's still an improvement over not considering the dependency at all.

> Also, it doesn't look like the fees of A, B, or C are counted, only the fees of the final child (D in this case).

IIRC this is intentional.

**Comment by laanwj** [comment]
Closing this for now. Should be reopened (or implemented otherwise) after the mempool refactor work.

**Comment by dcousens** [comment]
Maybe @laanwj tag it as blocked and keep it open?
I'm still not sure how this 'roadmap' is being tracked?

**Comment by luke-jr** [comment]
After 2 or 3 failed attempts to rewrite/port CPFP to 0.12, I am abandoning CPFP for the time being. Nobody seems to use it, and it's too complex.

Pretty sure the older versions also have a bug that they don't update grandchildren priority/feerate when their grandparents get mined. (Not a big deal.)

**Comment by schildbach** [comment]
@luke-jr Bitcoin Wallet users use it quite a bit. Current problem is only a few miners have implemented it.

**Comment by luke-jr** [comment]
Hmm, good to know. It's not going to make Knots 0.12.0 though - maybe I'll reconsider it later.

**Comment by rebroad** [comment]
Sorry if I'm being thick, I can see how this patch allows the recipient of a transaction to effectively pay the fee, but how often is this actually needed?

**Comment by sipa** [comment]
See #7600 for another implementation.

**Comment by schildbach** [comment]
@rebroad Especially in times like this week, with the network being very unreliable, CPFP is used a lot. Based on my logs, I'd say several thousand times a day.

**Comment by voisine** [comment]
Just wanted to add that we use CPFP in breadwallet as well, when spending unconfirmed inputs that don't come from the wallet's own change outputs. It's needed to get transactions unstuck if a user tries to spend their entire wallet balance including recent unconfirmed low/no-fee inputs.

**Comment by schildbach** [comment]
@voisine I assume in this case, the CPFP transaction is the same as the (presumably big) transaction that empties the wallet? Or are you somehow splitting it up into txns that act as CPFP and a transaction that empties the wallet?

**Comment by voisine** [comment]
this is if the user tries to spend more funds in a single tx than they have confirmed utxos that can satisfy

**Pull request closed without merging**